plugins/newrelic/src/api/index.ts

```diff
@@ -71,39 +76,105 @@ export interface NewRelicApi {
   getApplications(): Promise<NewRelicApplications>;
 }
 
+interface PaginationInformation {
+  nextPageLink: string;
+  relation: string;
+}
+
+interface NewRelicPageReadResult {
+  hasMoreApplications: boolean;
+  pagination: PaginationInformation | undefined;
+  applicationsFromReadPage: NewRelicApplication[];
+}
+
 export class NewRelicClient implements NewRelicApi {
   private readonly discoveryApi: DiscoveryApi;
+  private readonly fetchApi: FetchApi;
   private readonly proxyPathBase: string;
+  private baseUrl: string;
 
   constructor(options: Options) {
     this.discoveryApi = options.discoveryApi;
+    this.fetchApi = options.fetchApi;
     this.proxyPathBase = options.proxyPathBase ?? DEFAULT_PROXY_PATH_BASE;
+    this.baseUrl = '';
   }
 
   async getApplications(): Promise<NewRelicApplications> {
-    const url = await this.getApiUrl('apm', 'applications.json');
-    const response = await fetch(url);
-    let responseJson;
+    if (!this.baseUrl) {
+      const proxyUrl = await this.discoveryApi.getBaseUrl('proxy');
+      this.baseUrl = `${proxyUrl}${this.proxyPathBase}/apm/api/applications.json`;
+    }
+
+    let targetUrl = this.baseUrl;
+    let hasNextPage = true;
 
     try {
-      responseJson = await response.json();
+      let applications: NewRelicApplication[] = [];
+
+      do {
+        const { hasMoreApplications, pagination, applicationsFromReadPage } =
+          await this.fetchNewRelic(targetUrl);
+
+        hasNextPage = hasMoreApplications;
+        targetUrl = hasNextPage ? pagination!.nextPageLink : '';
+        applications = applications.concat(applicationsFromReadPage);
```

Comment:
you can declare the variable as const, and do `applications.concat(... applicationsFromReadPage)` if you like.

Reply:
Unfortunately, ```.concat``` does not modify any arrays. It only returns a new array with the two combined (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/concat):

![image](https://github.com/backstage/backstage/assets/129326788/888e7269-c44a-4acb-ab38-c9fde66a46cd)

Reply:
oh sorry that was a brain fart - i meant `.push(...`; push supports multiple arguments

Reply:
Switched to ```.push``` 👍